### PR TITLE
Bug 1467472 - Optimize IFV queries

### PR DESCRIPTION
**Pull request by sarah-clements**
* Remove get_repository helper and move functionality into `FailuresQuerySet` (to remove the 'all' filter, same as with the Commenter).
* Create custom paginator class to override django Paginator count method (this was causing the full table scan in `Failures` and `FailureCount`). DRF's OffsetLimitPagination class wasn't a good alternative to the PageNumberPagination class because it basically does the same thing with the count() method.
* Remove unnecessary order by in `FailureCount`.

These changes have resulted in a ~1 sec decrease in the queries. Since Ed filed a separate [bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1452311) for Python methods that should be updated, I'll do that in a separate pr.

### CODE REVIEW

File: treeherder/webapp/api/utils.py [hunk 1]
```diff
@@ -4,12 +4,14 @@
...
+# firefox-releases: mozilla-beta, mozilla-release
+# comm-releases: comm-beta
 REPO_GROUPS = {
-    'trunk': ['mozilla-central', 'mozilla-inbound', 'autoland'],
-    'firefox-releases': ['mozilla-beta', 'mozilla-release'],
-    'comm-releases': ['comm-beta', 'comm-release'],
+    'trunk': [1, 2, 77],
+    'firefox-releases': [6, 7],
+    'comm-releases': [38],
 }
```
Comment (sarah-clements) [inline treeherder/webapp/api/utils.py]: I couldn't find 'comm-release' in the repository table. Was `comm-releases` supposed to include comm-central or some other comm branch instead? If not, I think it makes sense to delete this line.

Comment (edmorley) [reply treeherder/webapp/api/utils.py]: It's quite surprising that passing IDs directly makes such a difference.

Could you give an example SQL statement for both (use eg django-debug-toolbar to output) so we can see there isn't something else going on?

If we do want to pass IDs directly, then I don't think hardcoding them here is the right thing to do. We could fetch them from the DB as an initial query before passing that to the main query.

Re comm-releases, good spot. Strangely the repository does exist (https://hg.mozilla.org/releases/comm-release/), it's just not been added to Treeherder. Perhaps they just haven't realised that they could ask for it to be added?

**Comment by sarah-clements** [comment]
@edmorley, I used the django-extensions shell plus since its easier to copy/paste:

>>> BugJobMap.objects.select_related('push', 'job', 'repository').filter(job__repository__name__in=['mozilla-central', 'mozilla-inbound', 'autoland'], job__push__time__range=('2018-09-05', '2018-09-12')).values('bug_id').annotate(bug_count=Count('job_id')).values('bug_id', 'bug_count').order_by('-bug_count')

SELECT `bug_job_map`.`bug_id`,
COUNT(`bug_job_map`.`job_id`) AS `bug_count`
FROM `bug_job_map` INNER JOIN `job`
ON (`bug_job_map`.`job_id` = `job`.`id`)
INNER JOIN `push`
ON (`job`.`push_id` = `push`.`id`)
INNER JOIN `repository`
ON (`job`.`repository_id` = `repository`.`id`)
WHERE (`push`.`time` BETWEEN '2018-09-05 00:00:00' AND '2018-09-12 00:00:00' AND `repository`.`name` IN ('mozilla-inbound', 'autoland', 'mozilla-central'))
 GROUP BY `bug_job_map`.`bug_id`
 ORDER BY `bug_count` DESC
 LIMIT 21

Execution time: 37.575178s [Database: default]

>>> BugJobMap.objects.select_related('push', 'job').filter(job__repository__id__in=[1, 2, 77], job__push__time__range=('2018-09-05', '2018-09-12')).values('bug_id').annotate(bug_count=Count('job_id')).values('bug_id', 'bug_count').order_by('-bug_count')

SELECT `bug_job_map`.`bug_id`,
COUNT(`bug_job_map`.`job_id`) AS `bug_count`
FROM `bug_job_map`
INNER JOIN `job`
ON (`bug_job_map`.`job_id` = `job`.`id`)
INNER JOIN `push`
ON (`job`.`push_id` = `push`.`id`)
WHERE (`job`.`repository_id` IN (1, 2, 77) AND `push`.`time` BETWEEN '2018-09-05 00:00:00' AND '2018-09-12 00:00:00')
GROUP BY `bug_job_map`.`bug_id`
ORDER BY `bug_count` DESC
LIMIT 21

Execution time: 1.494014s [Database: default]

Why would it be better to retrieve the IDs vs hard coding them? (happy to change it but I'm curious about the 'why') :)

**Comment by edmorley** [comment]
Thank you for those queries. I've tried playing around with them to make them faster but didn't stumble into a solution. At this point it's probably not worth trying any more and just going with the direct ID lookups.

Re fetching job IDs in an earlier query vs hardcoding, it's mostly to ensure they don't get out of sync (the ID is somewhat of an implementation detail) - however given that we have the IDs fixed in `repositories.json` I guess it could be worse. So long as there is a comment next to each saying what repository it's for (both to help the reader, plus also so anyone in the future deleting a repository finds those entries and deletes them too).

**Comment by sarah-clements** [comment]
@edmorley Ok, is there anything that needs to be changed? (There is already a comment above REPO_GROUPs stating the repo names for the ID's). About `comm-release`, who should I ask about that (if they want to add that repo to TH)?

**Comment by edmorley** [comment]
Ah yes there is a comment :-)

> About `comm-release`, who should I ask about that (if they want to add that repo to TH)?

Perhaps email Patrick? (most recent person to push to it, https://hg.mozilla.org/releases/comm-release/pushloghtml)

**Comment by camd** [comment]
> however given that we have the IDs fixed in repositories.json I guess it could be worse. 

OK, fair enough. I'll yield to Ed's wisdom here.  :)

**Pull request merged**